File: example/learning-agent/java/src/main/java/org/aiddl/example/learning_agent/functions/StateTransitionDataExtractor.java
package org.aiddl.example.learning_agent.functions;

import java.util.ArrayList;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

import org.aiddl.common.java.learning.LearningTerm;
import org.aiddl.common.java.planning.PlanningTerm;
import org.aiddl.core.java.interfaces.ConfigurableFunction;
import org.aiddl.core.java.representation.ListTerm;
import org.aiddl.core.java.representation.SetTerm;
import org.aiddl.core.java.representation.Term;
import org.aiddl.core.java.representation.TupleTerm;
import org.aiddl.core.java.function.FunctionRegistry;
import org.aiddl.core.java.tools.Logger;

public class StateTransitionDataExtractor implements ConfigurableFunction {

	private String name = StateTransitionDataExtractor.class.getSimpleName();
	boolean verbose = true;
		
	@Override
	public void configure(Map<Term, Term> settings, FunctionRegistry fReg) {
		this.verbose = Term.sym("true").equals(settings.get(Term.sym("verbose")));
	}

	@Override
	public Term apply( Term args ) {
		if ( verbose ) Logger.msg(this.getClass().getSimpleName(), "Extracting Data...");
		
		ListTerm attributes = (ListTerm) args.get(LearningTerm.Attributes);
		SetTerm sNextTerm = (SetTerm) args.get(Term.sym("next-state"));
		SetTerm sTerm = (SetTerm) args.get(PlanningTerm.State);
		Term actionTerm = args.get(Term.sym("action"));
		
		SetTerm currentData = (SetTerm) args.get(LearningTerm.Data);
		SetTerm updatedData = null;
		
		Set<Term> effects = new LinkedHashSet<>();
		for ( Term p : sNextTerm ) {
//			if ( sNextTerm.get(p.getKey()).equals(p.getValue())) {
				effects.add(p);
//			}
		}

//		Map<Term, Term> sTerm_sva = PlanningTerm.convert2svaMap(sTerm);
		List<Term> newDataPoint = new ArrayList<Term>();
		
		for ( int i = 0 ; i < attributes.size()-2 ; i++ ) {
			TupleTerm atom = (TupleTerm) attributes.get(i);
			newDataPoint.add(sTerm.get(atom));
		}
		newDataPoint.add(actionTerm);
		newDataPoint.add( Term.set(effects) );
		ListTerm newDataPointTerm = Term.list(newDataPoint);
		updatedData = currentData.add(newDataPointTerm);
		if ( verbose ) {
			if ( updatedData.size() != currentData.size() ) {
				Logger.msg(name, "New data point: " + newDataPoint);
			} else {
				Logger.msg(name, "Nothing new: " + newDataPoint);
			}
			
			Logger.incDepth();
			for ( Term d : updatedData ) {
				Logger.msg(name, d.toString());	
			}
			Logger.decDepth();					
			
			Logger.msg(name, String.format( "|Data| = %d", updatedData.size()) );
		}

		return updatedData;
	}
}

File: example/learning-agent-scala/src/main/scala/org/aiddl/example/learning_agent_scala/function/StateTransitionDataExtractor.scala
package org.aiddl.example.learning_agent_scala.function

import org.aiddl.core.scala.function.{Function, Verbose}
import org.aiddl.core.scala.representation.{Bool, CollectionTerm, KeyVal, ListTerm, Num, SetTerm, Sym, Term, Tuple, Var}
import org.aiddl.common.scala.learning.LearningTerm
import org.aiddl.common.scala.planning.PlanningTerm

import java.util.logging.Level

class StateTransitionDataExtractor(val attributes: ListTerm) extends Function with Verbose {
  this.logConfig(level=Level.INFO)

  override def apply(x: Term): Term = {
    this.logger.info("Extracting Data...")
    val state = x(PlanningTerm.State).asSet
    val action = x(Sym("action"))
    val nextState = x(Sym("next-state")).asSet
    val data = x(LearningTerm.Data).asList

    this(state, action, nextState, data)
  }

  def apply(state: Term, action: Term, nextState: SetTerm, data: ListTerm): ListTerm = {
    var updatedData: ListTerm = ListTerm.empty
    var effects: Set[Term] = Set.empty
    for (p <- nextState) {
      effects += p
    }
    var newDataPoint: List[Term] = Nil

    var i = 0
    while (i < attributes.size - 2) {
      val atom = attributes(i).asTup
      newDataPoint = state(atom) :: newDataPoint
      i += 1
    }
    newDataPoint = action :: newDataPoint
    newDataPoint = SetTerm(effects) :: newDataPoint

    val newDataPointTerm = ListTerm(newDataPoint.reverse)
    updatedData = data.add(newDataPointTerm)

    if (updatedData.size != data.size) {
      this.logger.info("New data point: " + newDataPointTerm)
    }
    this.logger.depth += 1
    for (d <- updatedData) {
      this.logger.fine(d.toString)
    }
    this.logger.depth -= 1
    this.logger.info(s"|Data| = ${updatedData.size}")
    updatedData
  }

}
